example/grammar/magnet.cpp

```diff
@@ -0,0 +1,496 @@
+            return t;
+        }
+    };
+
+    struct to_protocol {
+        string_view
+        operator()(query_param_encoded_view p)
+        {
+            url_view topic = parse_uri(p.value).value();
+            string_view t = topic.encoded_path();
+            std::size_t pos = t.find_last_of(':');
+            return t.substr(0, pos);
+        }
+    };
+
+public:
+    using topics_view =
+        filter_view<params_encoded_view, url_view, is_exact_topic, to_url>;
+
+    using infohashes_view =
+        filter_view<params_encoded_view, string_view, is_exact_topic, to_infohash>;
+
+    using protocols_view =
+        filter_view<params_encoded_view, string_view, is_exact_topic, to_protocol>;
+
+    using keys_view =
+        filter_view<params_encoded_view, pct_encoded_view, is_url_with_key, to_decoded_value>;
+
+    // URNs to the file or files hashes
+    // xt parameter or array [xt.1, xt.2, ...]
+    topics_view
+    exact_topics() const noexcept
+    {
+        return {u_.encoded_params()};
+    }
+
+    // info hash of the file or files
+    infohashes_view
+    info_hashes() const noexcept
+    {
+        return {u_.encoded_params()};
+    }
+
+    // protocols specified in the topics
+    protocols_view
+    protocols() const noexcept
+    {
+        return {u_.encoded_params()};
+    }
+
+    // Tracker URL; used to obtain resources for BitTorrent downloads
+    keys_view
+    address_trackers() const noexcept
+    {
+        return {
+            u_.encoded_params(),
+            is_url_with_key{"tr"}};
+    }
+
+    // direct download link
+    keys_view
+    exact_sources() const noexcept
+    {
+        return {
+            u_.encoded_params(),
+            is_url_with_key{"xs"}};
+    }
+
+    // fallback / direct download link
+    keys_view
+    acceptable_sources() const noexcept
+    {
+        return {
+            u_.encoded_params(),
+            is_url_with_key{"as"}};
+    }
+
+    // search keywords to search for in P2P networks
+    optional<pct_encoded_view>
+    keyword_topic() const noexcept
+    {
+        return decode_param("kt");
+    }
+
+    // Link to the metafile that contains a list of magneto
+    optional<pct_encoded_view>
+    manifest_topic() const noexcept
+    {
+        return decode_param("mt");
+    }
+
+    // A filename to display to the user
+    optional<pct_encoded_view>
+    display_name() const noexcept
+    {
+        return decode_param("dn");
+    }
+
+    // The payload data served over HTTP(S)
+    optional<pct_encoded_view>
+    web_seed() const noexcept
+    {
+        return decode_param("ws");
+    }
+
+    // informal options / parameters
+    optional<pct_encoded_view>
+    param(string_view key) const noexcept
+    {
+        params_encoded_view ps = u_.encoded_params();
+        auto it = ps.begin();
+        auto end = ps.end();
+        if (it != end)
+        {
+            query_param_encoded_view p = *it;
+            if (p.key.starts_with("x.") &&
```

Comment:
What if the key is percent-encoded?

Reply:
Yes. As in https://github.com/CPPAlliance/url/pull/262#discussion_r934722439, not having these helpers in pct_encoded_view makes it a little inconvenient. But I'll fix it and just do it the hard way.